File: source/algorithm/bfs.ts
type BfsNode<T> = { data: T; next: BfsNode<T> | null };
type BfsNodes<T> = BfsNode<T>[];

class Queue<T> {
    front: BfsNode<T> | null = null;
    rear: BfsNode<T> | null = null;
    #size = 0;

    get size() {
        return this.#size;
    }

    push(data: T) {
        const node: BfsNode<T> = {data, next: null};
        if (this.#size === 0) {
            this.front = node;
        } else {
            this.rear!.next = node;
        }
        this.rear = node;
        this.#size++;
    }

    pop(): undefined | T {
        if (this.#size === 0) {
            return undefined;
        }
        const node: BfsNode<T> = this.front!;
        this.front = node.next;
        this.#size--;
        return node.data;
    }

    isEmpty(): boolean {
        return this.size === 0;
    }

    peek(): T | undefined {
        if (this.#size === 0) {
            return undefined;
        }
        return this.front!.data;
    }
}

const bfs = (startIndex: number, nodes: BfsNodes<number>) => {
    const visit: { [index: number]: boolean } = {}
    const queue = new Queue<number>();
    queue.push(startIndex);
    visit[startIndex] = true;
    while (!queue.isEmpty()) {
        const x = queue.pop()!;
        console.log(x);
        let cur = nodes[x].next;
        while (cur) {
            let next = cur.data;
            if (!visit[next]) {
                queue.push(next);
                visit[next] = true;
            }
            cur = cur.next;
        }
    }
}

File: source/algorithm/bfs.c
//
// Created by 이병운 on 25. 5. 12.
//

#include <stdlib.h>
#include "bfs.h"
#include "../dataStructure/listStack.h"

void bfs(int index, Queue *queue, GraphNode **a) {
    queue->front = queue->rear = NULL;
    queue->size = 0;
    queuePush(index, queue);
    bfsVisit[index] = 1;
    while (queue->size != 0) {
        int x = queuePop(queue);
        printf("%d ", x);
        GraphNode *cur = a[x]->next;
        while (cur != NULL) {
            int next = cur->index;
            if (!bfsVisit[next]) {
                queuePush(index, queue);
                bfsVisit[next] = 1;
            }
            cur = cur->next;
        }
    }
}

void queuePush(int index, Queue *queue) {
    GraphNode *node = (GraphNode *) malloc(sizeof(GraphNode));
    node->index = index;
    node->next = NULL;
    if (queue->size == 0) {
        queue->front = node;
    } else {
        queue->rear->next = node;
    }
    queue->rear = node;
    queue->size++;
}

int queuePop(Queue *queue) {
    if (queue->size == 0) {
        return -INF;
    }
    GraphNode *node = queue->front;
    int index = node->index;
    queue->front = node->next;
    free(node);
    queue->size--;
    return index;
}